src/witness_gen.rs

```diff
@@ -39,6 +54,189 @@ impl<'a> WitnessGenerationData<'a> {
+                    .context("error converting local path to absolute path")?,
+                },
+            },
+        };
+
+        Ok(this)
+    }
+}
+
+impl WitnessDependency {
+    fn to_dependency_string(&self) -> String {
+        match &self.kind {
+            DependencyKind::Registry { version } => {
+                format!(r#"{{ version = "{version}", package = "{}"}}"#, self.name)
+            }
+            DependencyKind::Local { path } => format!(
+                r#"{{ path = "{}", package = "{}" }}"#,
+                path.display(),
+                self.name
+            ),
+        }
+    }
+}
+
+struct DependencyData {
+    target_dir: PathBuf,
+    baseline: WitnessDependency,
+    current: WitnessDependency,
+}
+
+impl DependencyData {
+    fn parse_input(witness_data: WitnessGenerationData) -> Result<Self> {
+        let baseline_data = witness_data
+            .baseline
+            .context("error parsing witness data, missing baseline crate data")?;
+        let current_data = witness_data
+            .current
+            .context("error parsing witness data, missing current crate data")?;
+        let target_dir = witness_data
+            .target_dir
+            .map(|path| path.to_path_buf())
+            .context("error parsing witness data, missing target directory")?;
+
+        let baseline = baseline_data
+            .try_into()
+            .context("error parsing baseline dependency")?;
+        let current = current_data
+            .try_into()
+            .context("error parsing baseline dependency")?;
+
+        Ok(Self {
+            target_dir,
+            baseline,
+            current,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct WitnessReport {
+    /// Total amount of time to run witness checks
+    time_elapsed: Duration,
+    /// Total number of lints with witness checks
+    lints_with_checks: usize,
+    /// Total number of witnesses generated and run
+    generated_checks: usize,
+    /// Total number of errored witnesses
+    errored_checks: usize,
+    /// Total number of failed checks
+    failed_checks: usize,
+    /// Total number of succeeded witnesses
+    succeeded_checks: usize,
+}
+
+#[derive(Debug)]
+pub(crate) struct WitnessResult {
+    pub check_results: Vec<Result<WitnessCheckResult>>,
+    pub breaking_change_found: bool,
+}
+
+#[derive(Debug)]
+pub(crate) struct WitnessCheckResult {
+    pub status: WitnessCheckStatus,
+    pub logic_result: Option<WitnessLogicResult>,
+}
+
+#[derive(Debug)]
+pub(crate) enum WitnessCheckStatus {
+    /// Indicates that `baseline` checked but `current` did not, which is the expected result
+    BreakingChange,
+
+    /// Indicated that the witness always checked successfully.
+    NoBreakingChange,
+
+    /// Indicates that the witness never checked successfully
+    NeverCompiled,
+
+    /// Indicates the unexpected result of `current` checking successfully, while `baseline` does not
+    InvertedBreak,
```

Comment:
If we checked `baseline` and it didn't compile, would we ever check `current` as opposed to just report a buggy witness? I'm not sure if seeing an `InvertedBreak` status would give us any additional actionable bits of information over `NeverCompiled`.